lib/elixir/lib/calendar.ex

```diff
@@ -1867,11 +2163,140 @@ defmodule DateTime do
+      |> apply_tz_offset(std_offset1)
+    {days2, {parts2, ppd2}} =
+      to_rata_die(datetime2)
+      |> apply_tz_offset(utc_offset2)
+      |> apply_tz_offset(std_offset2)
+
+    diff_days = days1 - days2
+    diff_ppd = ppd1 * ppd2
+    diff_parts = parts1 * ppd2 - parts2 * ppd1
+
+    # Keep integers in day fraction low.
+    gcd = gcd(diff_parts, diff_ppd)
+    diff_parts = div(diff_parts, gcd)
+    diff_ppd = div(diff_ppd, gcd)
+
+    if diff_parts < 0 do
+      {diff_days - 1, {diff_ppd + diff_parts, diff_ppd}}
+    else
+      {diff_days, {diff_parts, diff_ppd}}
+    end
+  end
+
+  @doc """
+  Converts a DateTime from one calendar to another.
+
+  ## Examples
+
+  iex> dt1 = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+  ...>                 hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+  ...>                 utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
+  iex> DateTime.convert(dt1, Calendar.ISO)
+  {:ok, %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+                   hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+                   utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}}
+  """
+  # TODO: For better examples and doctests we need a second (simple) calendar.
+  @spec convert(DateTime.t, Calendar.calendar) :: DateTime.t
+  # No conversion required if already of the target calendar.
+  def convert(%DateTime{calendar: calendar} = datetime, calendar) do
+    {:ok, datetime}
+  end
+
+  def convert(%DateTime{time_zone: time_zone, zone_abbr: zone_abbr, utc_offset: utc_offset, std_offset: std_offset} = datetime, calendar) do
+    result_datetime =
+      datetime
+      |> to_rata_die
+      |> from_rata_die(calendar)
+    {:ok, %DateTime{result_datetime | time_zone: time_zone, zone_abbr: zone_abbr, utc_offset: utc_offset, std_offset: std_offset}}
+  end
+
+
+  @spec convert!(DateTime.t, Calendar.calendar) :: DateTime.t
+  def convert!(datetime, calendar) do
+    case convert(datetime, calendar) do
+      {:ok, value} ->
+        value
+      {:error, reason} ->
+        raise ArgumentError, "cannot convert #{inspect datetime} to target calendar #{calendar}, reason: #{inspect reason}"
+    end
+  end
+
+  @spec to_rata_die(DateTime.t) :: Calendar.rata_die
+  defp to_rata_die(
+    %DateTime{calendar: calendar,
+              year: year, month: month, day: day, hour: hour, minute: minute, second: second, microsecond: microsecond,
+              time_zone: time_zone, zone_abbr: zone_abbr, utc_offset: utc_offset, std_offset: std_offset}
+  ) do
+    calendar.naive_datetime_to_rata_die(year, month, day, hour, minute, second, microsecond)
+  end
+
+  defp from_rata_die(rata_die, calendar) do
+    {year, month, day, hour, minute, second, microsecond} = calendar.naive_datetime_from_rata_die(rata_die)
+  end
+
+  # Integer representing seconds in the ISO 6801 calendar.
+  defp apply_tz_offset(rata_die, seconds) when is_integer(seconds) do
```

Comment:
We are not going to allow timezones in the day fraction format. If we have to, we will cross that bridge when we have to, but not before. :)